numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1074,33 +972,100 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
         Py_INCREF(Py_NotImplemented);
         return Py_NotImplemented;
     }
-    if (result == NULL) {
+
+    /*
+     * At this point we are locked into comparing (both are arrays) or
+     * something we would have tried to convert.
+     * If we are not in `==` and `!=`, this is an error and we hope that
```

Comment:
Is this not a change in behaviour for flexible `dtype`? Previously, that would return `NotImplemented` and give a `TypeError`. But I think a clearer error would be a win, so this is really just to check it is intended.

Reply:
I think we are fine?  We have strings supported in the ufunc now and the void dtypes are explicitly handled.  There are no further flexible dtypes left, since users effectively cannot create them.

Reply:
Indeed! Which is very nice!